csrc/multidevice/communication.cpp

```diff
@@ -138,6 +281,63 @@ c10::intrusive_ptr<c10d::Work> Broadcast::post(
 }
 
 Gather::Gather(CommParams params) : Communication(params, "gather") {
+  validateParams();
+}
+
+Gather::Gather(
+    TensorView* input_tv,
+    TensorView* output_tv,
+    at::Tensor input,
+    at::Tensor output,
+    DeviceIdxType my_device_index,
+    DeviceIdxType root)
+    : Communication("gather") {
+  params_.root = root;
+  DeviceMesh mesh = input_tv->getDeviceMesh();
+  params_.team = mesh.vector();
+  bool is_root_in_mesh = mesh.has(root);
+  if (!is_root_in_mesh) {
+    params_.team.push_back(root);
+  }
+
+  if (my_device_index == root &&
+      requiresRelayoutOutputTensor(input_tv, output_tv)) {
+    output_ = output;
+    output = allocateOutputTensor(input_tv, output);
+  }
+
+  if (requiresRelayoutOutputTensor(input_tv, output_tv) &&
+      mesh.has(my_device_index)) {
+    // Permute the device axis to the front.
+    // Input tensors are not copied since the device axis are size 1.
+    params_.src_bufs = {input.permute(permuteOrder(input_tv))};
```

Comment:
The data might not be ready when we instantiate the collective. This step should be done in "post"
Recall also that the same class instance could be reuse for multiple "post"s

Reply:
Are we sure that `permute` copies data and not just return a Tensor handle with changed meta data ?

Reply:
Good catch for post. 
Pytorch docs say permute changes the view, so it should not copy. It is a bit confusing though... all that is happening is a tensor that is [4, 3, DIDx(1)] needs to be viewed as [DIDx(1), 4, 3]. I think making this `view` will make it clear that there are no copies.